once more unto the decomposition breach, dear friends!

Sacha0 opened the pull request:
...or close up the wall with our obscure initialisms!

This pull request is take three on the recent decompositions rework; take two was #27159, and take one was #26997.

Specifically, this pull request deprecates `lufact[!]`, `schurfact[!]`, `lqfact[!]`, `qrfact[!]`, `ldltfact[!]`, and `svdfact[!]` to `lu[!]`, `schur[!]`, `lq[!]`, `qr[!]`, `ldlt[!]`, and `svd[!]` respectively, clobbering existing methods of the latter functions.

Additionally, this pull request (will eventually) deprecate `eig`/`eigfact[!]` to `eigen[!]`, `chol[!]`/`cholfact[!]` to `cholesky[!]`, `bkfact[!]` to `bunchkaufman[!]`, and `hessfact[!]` to `hessenberg[!]`.

To facilitate this deprecation, this pull request adds destructuring via iteration to some of the decomposition objects, and (immediately deprecated) `getindex(S::($DecompositionType), i::Integer)` methods that yield the decomposition components produced by iteration.

Pending items:
- [x] Deprecate `bkfact[!]` to `bunchkaufman[!]`
- [x] Deprecate `hessfact[!]` to `hessenberg[!]`
- [X] Deprecate `eigfact[!]` to `eigen[!]`
- [X] Deprecate `cholfact[!]` to `cholesky[!]`
- [X] Deprecate `eig` to `eigen`
- [ ] Deprecate `chol[!]` to `cholesky[!]`
- [X] Deprecate transitional `getindex` methods

Best!

Sacha0 committed:
Deprecate lufact to lu.

Sacha0 committed:
Deprecate schurfact to schur.

Sacha0 committed:
Deprecate lqfact to lq.

Sacha0 committed:
Deprecate qrfact to qr.

Sacha0 committed:
Deprecate ldltfact to ldlt.

Sacha0 committed:
Deprecate lufact! to lu!.

Sacha0 committed:
Deprecate schurfact! to schur!.

Sacha0 committed:
Deprecate lqfact! to lq!.

Sacha0 committed:
Deprecate qrfact! to qr!.

Sacha0 committed:
Deprecate ldltfact! to ldlt!.

Sacha0 committed:
Deprecate svdfact! to svd!.

Sacha0 committed:
Deprecate svdfact to svd.

Sacha0 committed:
Deprecate bkfact to bunchkaufman.

Sacha0 committed:
Deprecate bkfact! to bunchkaufman!.

Sacha0 committed:
Deprecate hessfact to hessenberg.

Sacha0 committed:
Deprecate hessfact! to hessenberg!.

Sacha0 committed:
Deprecate eigfact to eigen.

Sacha0 committed:
Deprecate eigfact! to eigen!.

Sacha0 committed:
Deprecate cholfact to cholesky.

Sacha0 committed:
Deprecate cholfact! to cholesky!.

Sacha0 committed:
Deprecate eig to eigen and destructuring via iteration.

Sacha0 committed:
Deprecate decomposition getindex methods in favor of property accessors.

Sacha0 committed:
Expand pull request numbers for some news items.

Sacha0 commented:
To my surprise and delight, I was able to get everything done this evening but the gentler `chol[!]` -> `cholesky[!]` deprecation. I hope to eke out time for any issues CI catches tomorrow morning, and will try for the preceding last deprecation then as well. If I don't manage that last commit prior to other obligations, I will try to eke out additional time between those obligations, but otherwise pass the torch to `@mbauman` till Thursday :).

fredrikekre commented:
Feel free to cherry pick from https://github.com/JuliaLang/julia/commits/breakmore

Sacha0 commented:
Incorporated Matt's `getindex` `@inline` removals into the relevant earlier commit, fixed the one CI failure in the `SparseArrays` suite, and removed the errant whitespace Matt found, so all commits up to "Expand pull request numbers for some news items" should be passing CI now.

Had a brief look at one of Fred's commits: I am not certain that `sqrt` is generally substitutable for `chol` in `choltype`, hence in the relevant commit of #27159 I kept `chol[!]` around internally as `legacychol[!]` for the moment. Might be overly conservative though :).

Out of time now, hope to find bits of time again in a couple of hours. Thanks all and best!

Sacha0 commented on stdlib/LinearAlgebra/src/deprecated.jl:
```diff
@@ -1,4 +1,4 @@
-# This file is a part of Julia. License is MIT: https://julialang.org/license
+ # This file is a part of Julia. License is MIT: https://julialang.org/license
```
Good catch! Should be fixed :).

Sacha0 commented on stdlib/LinearAlgebra/src/deprecated.jl:
```diff
@@ -7,6 +7,7 @@ using Base: @deprecate, depwarn
 @deprecate cond(F::LinearAlgebra.LU, p::Integer) cond(convert(AbstractArray, F), p)
 
 # PR #22188
+export cholfact, cholfact!
```
Indeed so :). I had forgotten at first, and once I remembered halfway through I decided to proceed consistently. Please feel welcome to remove the explicit exports!

mbauman replied:
Not a major concern by any means — just thought I'd mention it as I was reviewing.

simonbyrne commented:
It should be possible to deprecate `chol(X)` to `cholesky(X).U`

fredrikekre commented:
> I am not certain that `sqrt` is generally substitutable for `chol` in `choltype`

Maybe not, but we should only have to keep the legacy_chol method for `Number`, since `choltype` calls `promote_type(..., Float32)` which only makes sense for `Number`s (right?). I think it boils down to redefining `choltype` as:
```
function choltype(A)
    x = one(eltype(A))
    promote_type(typeof(x), typeof(sqrt(abs(real(x)))), Float32)
end
```
and then we don't need `chol` or `legacychol`.

> It should be possible to deprecate `chol(X)` to `cholesky(X).U`

62e0b7e75c1008120856955c2665e90b80377475

StefanKarpinski commented:
Failures are related to Pkg3 being renamed and/or timeouts.

Jutho commented:
I very much like the new factorization interface and all the work that has been done here. Congrats to everbody involved. 

I do have one question. I tried to look for a discussion about the name change `eig` to `eigen`. Why was this decided or deemed necessary? `svd` is also a three letter word so that can be it. I like the name changes from `chol` to `cholesky` and the full name `hessenberg`, but `eig` seems so standard. It's also used by Matlab and Numpy. In fact, if it were not for `eigen`, the names would be identical to Numpy/Scipy: `numpy.linalg.eig`, `numpy.linalg.svd`, `numpy.linalg.cholesky`, `scipy.linalg.hessenberg`, `scipy.linalg.schur`.

nalimilan commented:
See also https://discourse.julialang.org/t/eigen-the-odd-duck/14451.

KristofferC commented:
There are still no docs for `iterate` on factorization objects?
